src/coreclr/src/jit/hwintrinsicxarch.cpp

```diff
@@ -528,6 +528,118 @@ GenTree* Compiler::impBaseIntrinsic(NamedIntrinsic        intrinsic,
             break;
         }
 
+        case NI_Vector128_AsVector:
+        {
+            assert(sig->numArgs == 1);
+
+            if (getSIMDVectorRegisterByteLength() == YMM_REGSIZE_BYTES)
+            {
+                // Vector<T> is 32-bytes, so we should treat this as a call to Vector128.ToVector256
+                return impBaseIntrinsic(NI_Vector128_ToVector256, clsHnd, method, sig, mustExpand);
+            }
+
+            assert(getSIMDVectorRegisterByteLength() == XMM_REGSIZE_BYTES);
+
+            // We fold away the cast here, as it only exists to satisfy
+            // the type system. It is safe to do this here since the retNode type
+            // and the signature return type are both the same TYP_SIMD.
+
+            retNode = impSIMDPopStack(retType);
+            SetOpLclRelatedToSIMDIntrinsic(retNode);
+            assert(retNode->gtType == getSIMDTypeForSize(getSIMDTypeSizeInBytes(sig->retTypeSigClass)));
+
+            break;
+        }
+
+        case NI_Vector128_AsVector2:
+        case NI_Vector128_AsVector3:
+        case NI_Vector128_AsVector4:
+        {
+            // We fold away the cast here, as it only exists to satisfy
+            // the type system. It is safe to do this here since the retNode type
+            // and the signature return type are both the same TYP_SIMD or the
+            // return type is a smaller TYP_SIMD that shares the same register.
+
+            retNode = impSIMDPopStack(retType);
+            SetOpLclRelatedToSIMDIntrinsic(retNode);
+            assert(retNode->gtType == getSIMDTypeForSize(getSIMDTypeSizeInBytes(sig->retTypeSigClass)));
+
+            break;
+        }
+
+        case NI_Vector128_AsVector128:
+        {
+            assert(sig->numArgs == 1);
+
+            switch (getSIMDTypeForSize(simdSize))
+            {
+                case TYP_SIMD8:
+                case TYP_SIMD12:
+                {
+                    break;
+                }
+
+                case TYP_SIMD16:
+                {
+                    // We fold away the cast here, as it only exists to satisfy
+                    // the type system. It is safe to do this here since the retNode type
+                    // and the signature return type are both the same TYP_SIMD.
+
+                    retNode = impSIMDPopStack(retType);
+                    SetOpLclRelatedToSIMDIntrinsic(retNode);
+                    assert(retNode->gtType == getSIMDTypeForSize(getSIMDTypeSizeInBytes(sig->retTypeSigClass)));
+
+                    break;
+                }
+
+                default:
+                {
+                    unreached();
+                }
+            }
+
+            break;
+        }
+
+        case NI_Vector256_AsVector:
+        case NI_Vector256_AsVector256:
```

Comment:
`AsVector256` (which is `Vector<T>` to `Vector256<T>`) is similar.

If we don't support AVX, we just go to the software fallback
If we support AVX2 (meaning `Vector<T>` is `32-bytes`) then it is a nop
If we only support `AVX` we just set the upper 128-bits to zero via `ToVector256` (likewise, we probably want an `Unsafe` API so users can get non-deterministic bits, as we've done for other apis).

Reply:
> users can get non-deterministic bits

What does that mean? Who needs non-deterministic bits? This isn't a random number generator.

Reply:
The intent is to provide an API which provides fast conversion and requires the user to manage the upper bits themselves.

Similarly to how we have both `ToVector256` and `ToVector256Unsafe`. The former explicitly zeros the upper 128-bits and the latter is an effective nop that just allows access to the upper 128-bits of the register.